ENIGMAsystem/SHELL/Universal_System/Resources/AssetArray.h

```diff
@@ -45,8 +45,27 @@ namespace enigma {
 template<typename T>
 class AssetArray {
  public:
+  // Custom iterator for looping over only the existing assets in the array.
+  class iterator {
+   public:
+    iterator(AssetArray& assets, int ind): assets(assets), ind(ind) {}
+    iterator operator++() {
+      do { ++ind; }
+      while (!assets.exists(ind) && size_t(ind) < assets.size());
+      return *this;
+    }
+    bool operator!=(const iterator& other) const { return ind != other.ind; }
+    std::pair<int, T&> operator*() const { return {ind, assets[ind]}; }
+   private:
+    AssetArray& assets;
+    int ind;
+  };
+
   AssetArray() {}
 
+  iterator begin() { return iterator(*this, 0); }

```

Comment:
`this->get(0)` might not exist, and you wanted your iterator to return existing resources.